### PR TITLE
feat: error resolution flow control without exceptions

### PR DESCRIPTION
**Pull request by aepfli**
So far, we used exceptions to handle our flow control, but Exceptions are costly. In the end, we enriched our
evaluation Details with errorCode and errorMessage. If desired, the providers can also handle this to reduce the execution footprint in erroneous cases, which don't have to be exceptions. Till now, we also converted those into exceptions and threw them, with this little code change, we would return already the details, and call the hooks, without exceptions being thrown.

For example, FlagNotFound is such a case; it can be expected and not exceptional. Hence, the exception flow might add too much overhead to performance-critical environments.

This implementation now allows the prover implementors to decide whether to support either or both.

### Follow-up Tasks

Most likely this is something we also want to do in all other sdks.

### CODE REVIEW

**Comment by aepfli** [comment]
I am not good with performance tests, but I tried the following locally:

```java

        IntStream.range( 0, 1000000 ).parallel().forEach(i -> {
            String key = "key" + i;
            c.getBooleanDetails(key, false);
        });
```
With the exceptions, it took more than 3s; with details, it took only 1.5s.

Remarks:
- I changed the implementation of the `AlwaysBrokenWithDetailsProvider` to set a different message all the time so that we never create the same object if the JIT somehow starts to optimize.
- I removed the asserts because they took most of the time, and i was only aiming for run speed.
- I ran those tests for the fun about 5 times, with close results.
- I know this is not the best performance test, but it gives a little bit of insights

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java [hunk 1]
```diff
@@ -132,7 +132,11 @@ private <T> FlagEvaluationDetails<T> evaluateFlag(FlagValueType type, String key
 
             details = FlagEvaluationDetails.from(providerEval, key);
             if (details.getErrorCode() != null) {
-                throw ExceptionUtils.instantiateErrorByErrorCode(details.getErrorCode(), details.getErrorMessage());
+                OpenFeatureError error = ExceptionUtils.instantiateErrorByErrorCode(
+                        details.getErrorCode(),
+                        details.getErrorMessage());
+                details.setValue(defaultValue);
```
Comment (guidobrei) [inline src/main/java/dev/openfeature/sdk/OpenFeatureClient.java:138]: The previous implementation also sets the reason on the FlagEvaluationDetails

Comment (toddbaert) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java:138]: Yep this is important. The spec dictates it should be set in error conditions. Regardless of what the provider sets, we should set the reason to ERROR.

**Review by toddbaert — CHANGES_REQUESTED** [review]
I agree with @guidobrei 's point [here](https://github.com/open-feature/java-sdk/pull/1095/files#r1751449602).

In addition, I'm wondering if we should remove stack traces from some or all Errors, in addition to the FlagNotFound.

I think for a few, like `ProviderNotReady` and `TypeMismatch` we can also remove the stack trace, especially since these will be frequently instantiated by the SDK here and it's very obvious what the problem is.

**Comment by aepfli** [comment]
> In addition, I'm wondering if we should remove stack traces from some or all Errors, in addition to the FlagNotFound.
> 
> I think for a few, like `ProviderNotReady` and `TypeMismatch` we can also remove the stack trace, especially since these will be frequently instantiated by the SDK here and it's very obvious what the problem is.

I added this via a parent class to the errors mentioned. This way, we can easily identify them later. Wdyt? (we can also try to achieve this with an interface and default methods)

File: src/main/java/dev/openfeature/sdk/OpenFeatureClient.java (hunk 1 above)
Comment (aepfli) [reply src/main/java/dev/openfeature/sdk/OpenFeatureClient.java]: added and also added to the tests for verification

**Comment by guidobrei** [comment]
> I think for a few, like `ProviderNotReady` and `TypeMismatch` we can also remove the stack trace, especially since these will be frequently instantiated by the SDK here and it's very obvious what the problem is.

Not sure for the `TypeMismatch`. If `client.getBooleanValue("flag-key", false);` is called from user code for a string flag, I'd like to see the root cause.

**Comment by toddbaert** [comment]
> > I think for a few, like `ProviderNotReady` and `TypeMismatch` we can also remove the stack trace, especially since these will be frequently instantiated by the SDK here and it's very obvious what the problem is.
> 
> Not sure for the `TypeMismatch`. If `client.getBooleanValue("flag-key", false);` is called from user code for a string flag, I'd like to see the root cause.

OK I'm fine with that.

**Review by toddbaert — APPROVED** [review]

**Comment by toddbaert** [comment]
Approved but you can make the `TypeMismatch` change @guidobrei suggestes.

**Comment by jarebudev** [comment]
> ~I am not good with performance tests, but I tried the following locally:~
> 
> ```java
>         IntStream.range( 0, 1000000 ).parallel().forEach(i -> {
>             String key = "key" + i;
>             c.getBooleanDetails(key, false);
>         });
> ```
> 
> ~With the exceptions, it took more than 3s; with details, it took only 1.5s.~
> 
> Remarks:
> 
> * I changed the implementation of the `AlwaysBrokenWithDetailsProvider` to set a different message all the time so that we never create the same object if the JIT somehow starts to optimize.
> * I removed the asserts because they took most of the time, and i was only aiming for run speed.
> * I ran those tests for the fun about 5 times, with close results.
> * I know this is not the best performance test, but it gives a little bit of insights
> 
> // EDIT: forget about my tests, i might have made a mistake - i cant reproduce this - hence that, those are bad tests.
> 
> // EDIT: https://www.baeldung.com/java-exceptions-performance is a nice comprehensive post about this

This has got me wondering if there is value in creating some JMH benchmarks for the sdk...?

**Review by Kavindu-Dodan — APPROVED** [review]

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**